### ISSUE TITLE
Add tracing capability to IPv6 flows (tcpv6, udpv6 and icmpv6)

### ISSUE THREAD

**Issue by rafaeldtinoco** [rafaeldtinoco/conntracker#1]
The tracing feature uses the same mechanism as the one used by libnetfilter_log library. In theory we could be using libnetfilter_log (nflog_bind_pf / nflog_bind_group, etc), instead of using libmnl directly BUT the library does not provide access to the feature:

* NFULNL_CFG_F_CONNTRACK

which brings the conntrack information for the trace netlink message (telling us to which flow that trace log is related to).
It looks like feature NFULNL_CFG_F_CONNTRACK is NOT working with IPv6. 

Despite having:

```
ip6tables -t raw -I PREROUTING 1 -m conntrack --ctstate NEW,ESTABLISHED,RELATED -j ACCEPT
ip6tables -t raw -I OUTPUT 1 -m conntrack --ctstate NEW,ESTABLISHED,RELATED -j ACCEPT
ip6tables -t raw -A PREROUTING -j TRACE
ip6tables -t raw -A OUTPUT-j TRACE
```

It looks like the netlink ulog messages are not coming with conntrack feature when coming from IPv6 traces. I need to investigate why (kernel issue ?)

**Comment by rafaeldtinoco** [comment]
When executing conntracker in a newer Ubuntu (where iptables-nft is the default one) then the trace feature does not seem to work fine. This can be seen when comparing the output of its execution:

```
$ cat /tmp/conntracker.log 
 TCPv4 [           0] src = 10.250.97.1 (port=1024) to dst = 10.250.97.135 (port=22) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           1] src = 10.8.8.6 (port=1024) to dst = 149.154.175.51 (port=80) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           2] src = 10.8.8.6 (port=1024) to dst = 149.154.175.51 (port=443) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           3] src = 10.8.8.6 (port=1024) to dst = 149.154.175.53 (port=80) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           4] src = 10.8.8.6 (port=1024) to dst = 149.154.175.53 (port=443) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           5] src = 10.8.8.6 (port=1024) to dst = 172.217.162.202 (port=443) (confirmed)
                                table: filter, chain: OUTPUT, type: policy, position: 1
 TCPv4 [           6] src = 192.168.100.203 (port=1024) to dst = 192.168.100.13 (port=8009) (confirmed)
```

with the output of xtables-monitor tool:

```
PACKET: 2 daa9775e OUT=internal SRC=10.250.97.1 DST=10.250.97.135 LEN=52 TOS=0x0 TTL=64 ID=17563DF SPORT=42266 DPORT=22 ACK 
 TRACE: 2 daa9775e raw:OUTPUT:return:
 TRACE: 2 daa9775e raw:OUTPUT:policy:ACCEPT 
 TRACE: 2 daa9775e mangle:OUTPUT:return:
 TRACE: 2 daa9775e mangle:OUTPUT:policy:ACCEPT 
 TRACE: 2 daa9775e filter:OUTPUT:return:
 TRACE: 2 daa9775e filter:OUTPUT:policy:ACCEPT 
 TRACE: 2 daa9775e security:OUTPUT:return:
 TRACE: 2 daa9775e security:OUTPUT:policy:ACCEPT 
 TRACE: 2 daa9775e mangle:POSTROUTING:return:
 TRACE: 2 daa9775e mangle:POSTROUTING:policy:ACCEPT 
 TRACE: 2 0ba40eec raw:OUTPUT:rule:0x27b:CONTINUE -4 -t raw -A OUTPUT -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
-4 -t raw -A OUTPUT -s 10.250.97.1/32 -d 10.250.97.135/32 -p tcp -m tcp --dport 22 -j TRACE
-4 -t raw -A OUTPUT -s 127.0.0.1/32 -d 127.0.0.1/32 -p udp -m udp --dport 53 -j TRACE
-4 -t raw -A OUTPUT -s 192.168.100.203/32 -d 192.168.100.251/32 -p udp -m udp --dport 53 -j TRACE
-4 -t raw -A OUTPUT -s 192.168.100.203/32 -d 192.168.100.13/32 -p tcp -m tcp --dport 8009 -j TRACE
-4 -t raw -A OUTPUT -s 10.8.8.6/32 -d 172.217.162.202/32 -p tcp -m tcp --dport 443 -j TRACE
-4 -t raw -A OUTPUT -s 192.168.100.128/32 -d 255.255.255.255/32 -p udp -m udp --dport 67 -j TRACE
-4 -t raw -A OUTPUT -s 192.168.100.128/32 -d 224.0.0.251/32 -p udp -m udp --dport 5353 -j TRACE
-4 -t raw -A OUTPUT -s 10.8.8.6/32 -d 149.154.175.51/32 -p tcp -m tcp --dport 443 -j TRACE
-4 -t raw -A OUTPUT -s 10.8.8.6/32 -d 149.154.175.53/32 -p tcp -m tcp --dport 443 -j TRACE
-4 -t raw -A OUTPUT -s 10.8.8.6/32 -d 149.154.175.53/32 -p tcp -m tcp --dport 80 -j TRACE
-4 -t raw -A OUTPUT -s 10.8.8.6/32 -d 149.154.175.51/32 -p tcp -m tcp --dport 80 -j TRACE
PACKET: 2 0ba40eec OUT=internal SRC=10.250.97.1 DST=10.250.97.135 LEN=52 TOS=0x0 TTL=64 ID=17564DF SPORT=42266 DPORT=22 ACK 
 TRACE: 2 0ba40eec raw:OUTPUT:return:
 TRACE: 2 0ba40eec raw:OUTPUT:policy:ACCEPT 
 TRACE: 2 0ba40eec mangle:OUTPUT:return:
 TRACE: 2 0ba40eec mangle:OUTPUT:policy:ACCEPT 
 TRACE: 2 0ba40eec filter:OUTPUT:return:
 TRACE: 2 0ba40eec filter:OUTPUT:policy:ACCEPT 
 TRACE: 2 0ba40eec security:OUTPUT:return:
 TRACE: 2 0ba40eec security:OUTPUT:policy:ACCEPT 
 TRACE: 2 0ba40eec mangle:POSTROUTING:return:
 TRACE: 2 0ba40eec mangle:POSTROUTING:policy:ACCEPT 
```
xtables-monitor tool uses nftable netlink communication (and not libnfnetlink ones).

**Comment by rafaeldtinoco** [comment]
**workaround for now**: Make iptables-legacy the default iptables.

**Comment by rafaeldtinoco** [comment]
Just sent this e-mail to netfilter mailing list:

```
Hello list, Ken-ichirou and Pablo,

I'm developing github.com/rafaeldtinoco/conntracker and, during my tests, I 
realized that the feature:

NFULNL_CFG_F_CONNTRACK

does not seem to be giving me conntrack feature for IPv6 (as it does with 
IPv4).I have checked xtables-monitor code and IPv6 tracing seems to be
working when using libnftnl.. but I want to maintain compatibility to old
distros when nf-tables is not available/used.

bug: https://github.com/rafaeldtinoco/conntracker/issues/1

TL;DR version is:

I have a conntrack NEW,ESTABLISHED rule for everything in raw (so all the flows 
are accounted). To each identified flow, listened through libnetfilter-
conntrack, I add a rule to trace its events.

I map conntrack <-> trace relation through NFULNL_CFG_F_CONNTRACK feature, this 
way I know exactly through which rules that flow passed.

Example:

 UDPv4 [  2] src = 10.250.91.1 (port=1024) to dst = 10.250.91.255 (port=57621)
	table: raw, chain: PREROUTING, type: rule, position: 1
        table: nat, chain: OUTPUT, type: policy, position: 1
        table: nat, chain: POSTROUTING, type: policy, position: 2
        table: filter, chain: INPUT, type: policy, position: 1
        table: filter, chain: OUTPUT, type: policy, position: 1

Unfortunately with:

nflog_nlmsg_parse(nlh, attrs), my attrs[NFULA_CT] is always NULL for IPv6:

 TCPv6 [  0] src = fe80::1453:5dff:fe1a:ca68 (port=1024) to dst = fe80::216:3eff:fe7f:aedd (port=22) (confirmed)
ICMPv6 [  0] src = fe80::1453:5dff:fe1a:ca68 to dst = fe80::216:3eff:fe7f:aedd (type=0 | code=0) (confirmed)
ICMPv6 [  1] src = fe80::1453:5dff:fe1a:ca68 to dst = fe80::3c76:fdff:fea2:82b4 (type=0 | code=0)

This can also be observed with:

libnetfilter-log/utils/nf-log.c code (if BUILD_NFCT)

which I used as a base for this part of the code.

Is there a reason why IPv6 conntrack info is not being sent through netlink ?

Thanks in advance!

-rafaeldtinoco
```
at: https://marc.info/?l=netfilter-devel&r=1&w=2
